### PR TITLE
refactor(logger): BYOFormatter and Handler, UTC support, and more

### PR DESCRIPTION
**Pull request by heitorlessa**
**Issue #, if available:** #381 #387 #215 #98

## Description of changes:

Major refactor of Logger to **a)** make it more flexible for those coming from other loggers, **b)** bring your own formatter for those with a common structure that differs from our opinionated one, **c)** UX enhancements to add or remove keys from the logger, **d)** address long standing tech debts on  `LambdaPowertoolsFormatter`.

### UTC support

> Note: By default, Python standard logging use local time. 

```python
from aws_lambda_powertools import Logger

logger = Logger(utc=True)

logger.info("Hello")  # "timestamp":"2021-04-19 12:58:46.635+0100", dst
```

### New convenience property to change logger handler, if necessary

```python
from aws_lambda_powertools import Logger

logger = Logger()

handler = logger.registered_handler  # takes into account parent or child loggers
```

### Bring your own handler

> Note: By default, we log to standard output due to Lambda integration. This unlocks using Logger in places like Glue, or locally when you want to log into a filename for example.

```python
from aws_lambda_powertools import Logger

handler = logging.FileHandler(filename="log_file.json")
logger = Logger(logger_handler=handler)

logger.info("Hello")  # populatest `log_file.json` in your filesystem
```

### Bring your own formatter

```python
from aws_lambda_powertools import Logger
from aws_lambda_powertools.logging.formatter import BasePowertoolsFormatter

class CustomFormatter(BasePowertoolsFormatter):
    custom_format = {}

    def append_keys(self, **additional_keys):
        # used by inject_lambda_context decorator, Logger initialization, and structure_log
        self.custom_format.update(additional_keys)

    def remove_keys(self, keys: Iterable[str]):
        # used by Logger public method
        for key in keys:
            self.custom_format.pop(key, None)

    def format(self, record: logging.LogRecord) -> str:  # noqa: A003
        return json.dumps(
            {
                "message": super().format(record),
                "timestamp": self.formatTime(record),
                "my_default_key": "test",
                **self.custom_format,
            }
        )

custom_formatter = CustomFormatter()
logger = Logger(logger_formatter=custom_formatter)
```

### Updating formatter made easier

```python
from aws_lambda_powertools import Logger

logger = Logger()

logger.append_keys(request_id="id", order_id="id")
logger.info("with both request and order id keys")

logger.remove_keys(["request_id"])
logger.info("with request id key only")
```

**Checklist**

 

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] Update tests
* [ ] Update docs
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)
* [x] **New features**
  * [x] Bring your own formatter via `Logger(log_formatter=...)`
    * [x] Create ABC to support custom formatters to enforce `append_keys` and `remove_keys`
    * [x] Update Logger to use `append_keys` for Lambda decorator and `structure_log` calls
  * [x] Add support for `utc=True`
  	* [x] Include timezone in `timestamp` key
  	* [x] Create custom msec string interpolation to not break previous format (RFC 3339), since std logging doesn't support msec when built-in `formatTime` is overriden
* [x] **Tech debt**
  * [x] Introduce `append_keys`, `remove_keys` over `structure_log(append=True)` to align with other runtimes
  * [x] Allow customers to set bare minimum (only message, and xray_trace_id)
    * [x] Allow to suppress sampling_rate
    * [x] Remove sampling_rate when not in use (set to None by default over 0.0)
  * [x] Allow custom JSON serializer/deserializer for those seeking faster performance
  * [x] Compact logging to save on ingestion & storage cost (`separators[]` to remove whitespace)
  * [x] [Internal] Refactor standalone methods into staticmethods
  * [x] Rename testing to powertools formatter given we have vastly deviated from the original implementation
    * [x] Create alias for JsonFormatter to prevent breaking existing customers
  * [x] Add convenience properties to access handler and formatter
  * [x] Support all standard log attrs (%()s) if given via kwargs
* [x] Support custom handler
* [x] Update docstring parameters in Logger
* [x] Update docstring parameters in Formatter

---

**Create separate PR for docs to include enhancements**:

* [ ] Document examples
  * [ ] Default JSON fn for unserializable (json.dumps(data, default=json_default))
  * [ ] Std log attrs %()s support
  * [ ] Bring your own formatter
  * [ ] Bare minimal attrs (level, message, service)
  * [ ] Update all docs with additional TZ in timestamp
  * [ ] Bringing `orjson` for those seeking to gain every possible microsecond
* [ ] Update `Testing your code` section to include how to use `stream=sys.stdout`

## Breaking change checklist

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by michaelbrewer** [comment]
@heitorlessa nice, this would help alot when you have company mandated structured logging formats.

**Comment by heitorlessa** [comment]
@bml1g12 - what are your thoughts on the UTC UX for your use case?

Anything else you need while I'm at it?

**Comment by bml1g12** [comment]
Currently, I am declaring in each handler 
```
LOGGER = Logger(datefmt="%Y-%m-%dT%H:%M:%SZ")
LOGGER._logger.handlers[0].formatter.converter = time.gmtime  #pylint: disable = protected-access
```
So in this new UX I would instead simply do
```
LOGGER = Logger(datefmt="%Y-%m-%dT%H:%M:%SZ", utc=True)
```
In each handler and the child would inherit this right?

If so, looks great to me! Much appreciated addition.

**Comment by heitorlessa** [comment]
Yes exactly. I noticed you’re using RFC 8601 in your datefmt, I’ll make
sure this will be the default timestamp in v2, as changing now would break
customers.

**Comment by bml1g12** [comment]
Yeah I chose this as I saw a few blogs recommending it as a standard across languages for structured logging; I think its the default for javascript logging library maybe

**Comment by michaelbrewer** [comment]
@heitorlessa I am liking these changes so long, I am glad to see that now even `sampling_rate` can be suppressed :)

File: aws_lambda_powertools/logging/formatter.py [hunk 1]
```diff
@@ -27,63 +31,140 @@
...
+        # so we create a custom one (%F) and replace logging record ts
+        # Reason 2 is that std logging doesn't support msec after TZ
+        record_ts = self.converter(record.created)
+        if datefmt:
+            ts_formatted = time.strftime(datefmt, record_ts)
+        else:
+            msec = "%03d" % record.msecs
+            custom_fmt = self.default_time_format.replace(self.custom_ms_time_directive, msec)
+            ts_formatted = time.strftime(custom_fmt, record_ts)
+        return ts_formatted
```
Comment (michaelbrewer) [inline aws_lambda_powertools/logging/formatter.py:144]:
```suggestion
        record_ts = self.converter(record.created)
        if datefmt:
            return time.strftime(datefmt, record_ts)

        # NOTE: Python `time.strftime` doesn't provide msec directives
        # so we create a custom one (%F) and replace logging record ts
        # Reason 2 is that std logging doesn't support msec after TZ
        msecs = "%03d" % record.msecs
        custom_fmt = self.default_time_format.replace(self.custom_ms_time_directive, msecs)
        return time.strftime(custom_fmt, record_ts)
```

**Review by michaelbrewer — CHANGES_REQUESTED** [review]
LGTM, just a minor tweak and typo

**Pull request merged**